File: regex-test/dotnet/TesterAsync.cs
using System;
using System.Text.RegularExpressions;

string line;
string testLine;

while ((line = Console.ReadLine()) != null) {
    if (!line.StartsWith("REGEX:")) {
        continue;
    }

    try {
        var r = new Regex(line.Substring(6), RegexOptions.Compiled);
        Console.WriteLine("success");

        while ((testLine = Console.ReadLine()) != null && testLine.StartsWith("TEST:")) {
            var test = testLine.Substring(5);
            if (r.IsMatch(test)) {
                Console.WriteLine("test good");
            } else {
                throw new ArgumentException($"Regex '{r}' does not match '{test}'");
            }
        }
    } catch (ArgumentException e) {
        string message = e.Message.Replace("\\", @"\\").Replace("\n", @"\n");
        Console.WriteLine(message);
    }
}

File: regex-test/js/tester-async.js
const process = require('node:process')

/** @type {RegExp|undefined} */
let regex

process.stdin.on('data', (data) => {
  let line = data.toString()
  if (line.endsWith('\r\n')) {
    line = line.slice(0, line.length - 2)
  } else if (line.endsWith('\n')) {
    line = line.slice(0, line.length - 1)
  }

  if (regex === undefined) {
    try {
      regex = new RegExp(line, 'u')
      console.log('success')
    } catch (e) {
      console.log(substituteLf(e.message))
    }
  } else if (line.startsWith('TEST:')) {
    const test = line.slice(5)

    if (regex.test(test)) {
      console.log('test good')
    } else {
      console.log(
        substituteLf(`Regex '${regex.source}' does not match '${test}'`)
      )
      regex = undefined
    }
  } else {
    regex = undefined
  }
})

function substituteLf(s = '') {
  return s.replace(/[\n\\]/g, (c) => (c === '\\' ? '\\\\' : '\\n'))
}
